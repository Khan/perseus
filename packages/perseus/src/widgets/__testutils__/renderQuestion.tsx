import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
