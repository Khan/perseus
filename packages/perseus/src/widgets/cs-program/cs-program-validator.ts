import type {PerseusScore} from "../../types";
import type {PerseusCSProgramUserInput} from "../../validation.types";

export function csProgramValidator(
    state: PerseusCSProgramUserInput,
): PerseusScore {
    // The iframe can tell us whether it's correct or incorrect,
    //  and pass an optional message
    if (state.status === "correct") {
        return {
            type: "points",
            earned: 1,
            total: 1,
            message: state.message || null,
        };
    }
    if (state.status === "incorrect") {
        return {
            type: "points",
            earned: 0,
            total: 1,
            message: state.message || null,
        };
    }
    return {
        type: "invalid",
        message: "Keep going, you're not there yet!",
    };
}
