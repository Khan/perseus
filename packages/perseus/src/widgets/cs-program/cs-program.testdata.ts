import type {PerseusRenderer} from "@khanacademy/perseus-core";

export const question1: PerseusRenderer = {
    content: "[[\u2603 cs-program 1]]\n\n",
    images: {},
    widgets: {
        "cs-program 1": {
            graded: true,
            version: {major: 0, minor: 0},
            static: false,
            type: "cs-program",
            options: {
                settings: [
                    {name: "", value: ""},
                    {name: "", value: ""},
                ],
                height: 540,
                programID: "6293105639817216",
                static: false,
                showButtons: false,
                showEditor: false,
            },
            alignment: "block",
        },
    },
};
