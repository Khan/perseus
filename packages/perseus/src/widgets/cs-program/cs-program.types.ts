export type Status = "correct" | "incorrect" | "incomplete";

export type UserInput = {
    status: Status;
    message: string | null;
};
