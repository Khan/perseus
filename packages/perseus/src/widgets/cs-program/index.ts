export {default} from "./cs-program";
