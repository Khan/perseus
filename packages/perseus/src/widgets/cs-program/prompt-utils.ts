import {WidgetType} from "../../prompt-types";

import type {PerseusCSProgramWidgetOptions} from "../../perseus-types";
import type {PerseusCSProgramUserInput} from "../../validation.types";

export type CSProgramPromptJSON = {
    type: WidgetType;
    programID: PerseusCSProgramWidgetOptions["programID"];
    userInput: {
        status: PerseusCSProgramUserInput["status"];
        message: PerseusCSProgramUserInput["message"];
    };
};

export const getPromptJSON = (
    renderProps: PerseusCSProgramWidgetOptions,
    userInput: PerseusCSProgramUserInput,
): CSProgramPromptJSON => {
    return {
        type: WidgetType.CS_PROGRAM,
        programID: renderProps.programID,
        userInput: {
            status: userInput.status,
            message: userInput.message,
        },
    };
};
