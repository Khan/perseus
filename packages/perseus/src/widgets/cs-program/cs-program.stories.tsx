import * as React from "react";

import {ServerItemRendererWithDebugUI} from "../../../../../testing/server-item-renderer-with-debug-ui";
import {generateTestPerseusItem} from "../../util/test-utils";

import {question1} from "./cs-program.testdata";

export default {
    title: "Perseus/Widgets/CS Program",
};

type StoryArgs = Record<any, any>;

export const Question1 = (args: StoryArgs): React.ReactElement => {
    return (
        <ServerItemRendererWithDebugUI
            item={generateTestPerseusItem({question: question1})}
        />
    );
};
