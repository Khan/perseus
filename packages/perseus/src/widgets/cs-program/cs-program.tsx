/**
 * This widget is for embedding Khan Academy CS programs.
 */

import {StyleSheet, css} from "aphrodite";
import $ from "jquery";
import * as React from "react";
import _ from "underscore";

import {getDependencies} from "../../dependencies";
import * as Changeable from "../../mixins/changeable";
import {articleMaxWidthInPx} from "../../styles/constants";
import Util from "../../util";
import {isFileProtocol} from "../../util/mobile-native-utils";
import {toAbsoluteUrl} from "../../util/url-utils";

import {csProgramValidator} from "./cs-program-validator";

import type {PerseusCSProgramWidgetOptions} from "../../perseus-types";
import type {Widget, WidgetExports, WidgetProps} from "../../types";
import type {
    PerseusCSProgramRubric,
    PerseusCSProgramUserInput,
} from "../../validation.types";

const {updateQueryString} = Util;

type RenderProps = PerseusCSProgramWidgetOptions & PerseusCSProgramUserInput;

type Props = WidgetProps<RenderProps, PerseusCSProgramRubric>;

type DefaultProps = {
    showEditor: Props["showEditor"];
    showButtons: Props["showButtons"];
    status: Props["status"];
    // optional message
    message: Props["message"];
};

function getUrlFromProgramID(programID: any) {
    const {InitialRequestUrl} = getDependencies();

    const path =
        "/computer-programming/program/" +
        `${programID}/embedded?embed=yes&author=no`;
    // When loading scratchpads in our native apps, the current URL will exist
    // as a `file://` URL. In this case, we want to reference `ka.org` instead
    // of defaulting to what we get from `toAbsoluteUrl`.
    if (isFileProtocol(InitialRequestUrl.protocol)) {
        return `https://khanacademy.org${path}`;
    }

    return toAbsoluteUrl(path);
}

/* This renders the scratchpad in an iframe and handles validation via
 * window.postMessage */
class CSProgram extends React.Component<Props> implements Widget {
    static defaultProps: DefaultProps = {
        showEditor: false,
        showButtons: false,
        status: "incomplete",
        // optional message
        message: null,
    };

    componentDidMount() {
        $(window).on("message", this.handleMessageEvent);
    }

    componentWillUnmount() {
        $(window).off("message", this.handleMessageEvent);
    }

    handleMessageEvent: (arg1: any) => void = (e) => {
        // We receive data from the iframe that contains {passed: true/false}
        //  and use that to set the status
        // It could also contain an optional message
        let data: Record<string, any> = {};
        try {
            data = JSON.parse(e.originalEvent.data);
        } catch (err: any) {
            return;
        }

        if (_.isUndefined(data.testsPassed)) {
            return;
        }

        const status = data.testsPassed ? "correct" : "incorrect";
        this.change({
            status: status,
            message: data.message,
        });
    };

    change: (...args: ReadonlyArray<unknown>) => any = (...args) => {
        // @ts-expect-error - TS2345 - Argument of type 'readonly unknown[]' is not assignable to parameter of type 'any[]'.
        return Changeable.change.apply(this, args);
    };

    getUserInput(): PerseusCSProgramUserInput {
        return {
            status: this.props.status,
            message: this.props.message,
        };
    }

    render(): React.ReactNode {
        if (!this.props.programID) {
            return <div />;
        }

        let styleContainer = false;
        let url = getUrlFromProgramID(this.props.programID);
        let className;
        const style = {
            height: this.props.height,
            width: "100%",
        } as const;

        if (this.props.showEditor) {
            url += "&editor=yes";
            className = "perseus-scratchpad-editor";
        } else {
            url += `&editor=no&width=${articleMaxWidthInPx}`;
            className = "perseus-scratchpad";
            if (this.props.programType !== "webpage") {
                styleContainer = true;
            }
        }

        if (this.props.showButtons) {
            url += "&buttons=yes";
            // Matches templates/scratchpads/embed_script.js
            // Toolbar height is 66, border height is 1 pixel
            // @ts-expect-error - TS2540 - Cannot assign to 'height' because it is a read-only property.
            style.height += 67;
        } else {
            url += "&buttons=no";
        }

        // Turn array of [{name: "", value: ""}] into object
        if (this.props.settings) {
            const settings: Record<string, any> = {};
            _.each(this.props.settings, function (setting) {
                if (setting.name && setting.value) {
                    settings[setting.name] = setting.value;
                }
            });
            // This becomes available to programs as Program.settings()
            url = updateQueryString(url, "settings", JSON.stringify(settings));
        }

        const sandboxOptions = [
            "allow-popups",
            "allow-same-origin",
            "allow-scripts",
            "allow-top-navigation",
        ].join(" ");

        // We sandbox the iframe so that we allowlist only the functionality
        //  that we need. This makes it a bit safer in case some content
        //  creator "went wild".
        // http://www.html5rocks.com/en/tutorials/security/sandboxed-iframes/
        return (
            <div
                className={css(
                    styles.widthOverride,
                    styleContainer && styles.container,
                )}
            >
                <iframe
                    sandbox={sandboxOptions}
                    src={url}
                    style={style}
                    className={className}
                    allowFullScreen={true}
                />
            </div>
        );
    }
}

const styles = StyleSheet.create({
    // Override the inherited width from the perseus paragraph class
    // 820 is minimum width for 420px editor and 400px canvas
    widthOverride: {
        width: 820,
    },
    container: {
        margin: "auto",
    },
});

export default {
    name: "cs-program",
    displayName: "CS Program",
    supportedAlignments: ["block", "full-width"],
    widget: CSProgram,
    hidden: true,
    validator: csProgramValidator,
} satisfies WidgetExports<typeof CSProgram>;
