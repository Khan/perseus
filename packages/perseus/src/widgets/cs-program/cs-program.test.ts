import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {renderQuestion} from "../__testutils__/renderQuestion";

import {question1} from "./cs-program.testdata";
import {type CSProgramPromptJSON} from "./prompt-utils";

import type {PerseusCSProgramUserInput} from "../../validation.types";

describe("cs-program widget", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should snapshot", () => {
        // Arrange
        const apiOptions = {
            isMobile: false,
        } as const;

        // Act
        const {container} = renderQuestion(question1, apiOptions);

        // Assert
        expect(container).toMatchSnapshot("first render");
    });

    it("should snapshot on mobile", () => {
        // Arrange
        const apiOptions = {
            isMobile: true,
        } as const;

        // Act
        const {container} = renderQuestion(question1, apiOptions);

        // Assert
        expect(container).toMatchSnapshot("first mobile render");
    });

    it("should show default user input before user interaction", () => {
        const apiOptions = {
            isMobile: false,
        } as const;

        const {renderer} = renderQuestion(question1, apiOptions);
        const userInput =
            renderer.getUserInput()[0] as PerseusCSProgramUserInput;

        expect(userInput.status).toBe("incomplete");
        expect(userInput.message).toBe(null);
    });

    it("Should get prompt json which matches the state of the UI", async () => {
        // Arrange
        const apiOptions = {
            isMobile: false,
        } as const;

        const {renderer} = renderQuestion(question1, apiOptions);

        const widget = renderer.getWidgetInstance("cs-program 1");

        // Act
        const json = widget?.getPromptJSON?.() as CSProgramPromptJSON;

        // Assert
        expect(json).toEqual({
            type: "cs-program",
            isSupported: false,
        });
    });
});
