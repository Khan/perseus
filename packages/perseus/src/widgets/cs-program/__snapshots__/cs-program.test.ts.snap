// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cs-program widget should snapshot on mobile: first mobile render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="widthOverride_1vdzrqd-o_O-container_1v2yujs"
          >
            <iframe
              allowfullscreen=""
              class="perseus-scratchpad"
              sandbox="allow-popups allow-same-origin allow-scripts allow-top-navigation"
              src="http://localhost:8081/computer-programming/program/6293105639817216/embedded?embed=yes&author=no&editor=no&width=688&buttons=no&settings=%7B%7D"
              style="height: 540px; width: 100%;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`cs-program widget should snapshot: first render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="widthOverride_1vdzrqd-o_O-container_1v2yujs"
          >
            <iframe
              allowfullscreen=""
              class="perseus-scratchpad"
              sandbox="allow-popups allow-same-origin allow-scripts allow-top-navigation"
              src="http://localhost:8081/computer-programming/program/6293105639817216/embedded?embed=yes&author=no&editor=no&width=688&buttons=no&settings=%7B%7D"
              style="height: 540px; width: 100%;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
